hnn_core/dipole.py

```diff
@@ -145,6 +145,101 @@ def simulate_dipole(net, n_trials=1, n_jobs=1):
     return dpl
 
 
+def average_dipoles(dpls):
+    """Compute average over a list of Dipole objects.
+    Parameters
+    ----------
+    dpls: list of Dipole objects
+        Contains list of dipole results to be averaged
+
+    Returns
+    -------
+    dpl: instance of Dipole
+        A dipole object with averages of the dipole data
+    """
+    # need at least on Dipole to get times
+    assert (len(dpls) > 0)
+
+    agg_avg = np.mean(np.array([dpl.dpl['agg'] for dpl in dpls]), axis=0)
+    L5_avg = np.mean(np.array([dpl.dpl['L5'] for dpl in dpls]), axis=0)
+    L2_avg = np.mean(np.array([dpl.dpl['L2'] for dpl in dpls]), axis=0)
+
+    avg_dpl_data = np.c_[agg_avg,
+                         L2_avg,
+                         L5_avg]
+
+    avg_dpl = Dipole(dpls[0].t, avg_dpl_data)
+
+    return avg_dpl
+
+
+def rmse(dpl, exp_dpl, tstart=0.0, tstop=0.0, weights=None):
+    """ Calculates RMSE between data in dpl and exp_dpl
+    Parameters
+    ----------
+    dpl: instance of Dipole
+        A dipole object with simulated data
+    exp_dpl: instance of Dipole
+        A dipole object with experimental data
+    tstart | None: float
+        Time at beginning of range over which to calculate RMSE
+    tstop | None: float
+        Time at end of range over which to calculate RMSE
+    weights | None: array
+        An array of weights to be applied to each point in
+        simulated dpl. Must have length >= dpl.dpl
+        If None, weights will be replaced with 1's for typical RMSE
+        calculation.
+
+    Returns
+    -------
+    err: float
+        Weighted RMSE between data in dpl and exp_dpl
+    """
+    from scipy import signal
+
+    exp_times = exp_dpl.t
+    sim_times = dpl.t
+
+    # do tstart and tstop fall within both datasets?
+    # if not, use the closest data point as the new tstop/tstart
+    for tseries in [exp_times, sim_times]:
+        if tstart < tseries[0]:
+            tstart = tseries[0]
+        if tstop > tseries[-1]:
+            tstop = tseries[-1]
+
+    # make sure start and end times are valid for both dipoles
+    exp_start_index = (np.abs(exp_times - tstart)).argmin()
+    exp_end_index = (np.abs(exp_times - tstop)).argmin()
+    exp_length = exp_end_index - exp_start_index
+
+    sim_start_index = (np.abs(sim_times - tstart)).argmin()
+    sim_end_index = (np.abs(sim_times - tstop)).argmin()
+    sim_length = sim_end_index - sim_start_index
+
+    if weights is None:
+        # weighted RMSE with weights of all 1's is equivalent to
+        # normal RMSE
+        weights = np.ones(len(sim_times[0:sim_end_index]))
+    weights = weights[sim_start_index:sim_end_index]
+
+    dpl1 = dpl.dpl['agg'][sim_start_index:sim_end_index]
+    dpl2 = exp_dpl.dpl['agg'][exp_start_index:exp_end_index]
+
+    if (sim_length > exp_length):
+        # downsample simulation timeseries to match exp data
```

Comment:
even this is dangerous. Why not simulate with the correct `sfreq` in the first place?

Reply:
@jasmainak We usually simulate at time steps of around 0.01 ms to maintain stability during numerical integration of the HH model. Since 100,000 Hz is very impractical for experimental measurements, we're forced to downsample the simulation time series. What is dangerous about it?

Reply:
If you don't have force the two signals to have equal number of samples, the differences in rmse could be because of: a) Different sampling frequency, b) Different length of signal. Now because all this is happening within one function, you will have no control over it. To me the `rmse` function should only accept signals of equal number of samples. Then if you want to crop or resample, there should be different methods ``dipole.crop`` and ``dipole.resample`` for it. And the time points are stored in ``dipole.times``. See how we do it in MNE: https://mne.tools/dev/generated/mne.Evoked.html#mne.Evoked.

Also, it's weird that you resample such a short signal, wouldn't that introduce ringing artifacts?

Reply:
I'd like to pick this back up with a rebase. Though, there are some discussions to follow up on.

> If you don't have force the two signals to have equal number of samples, the differences in rmse could be because of: a) Different sampling frequency, b) Different length of signal. Now because all this is happening within one function, you will have no control over it. To me the `rmse` function should only accept signals of equal number of samples. Then if you want to crop or resample, there should be different methods `dipole.crop` and `dipole.resample` for it. And the time points are stored in `dipole.times`. See how we do it in MNE: https://mne.tools/dev/generated/mne.Evoked.html#mne.Evoked.
> 
> Also, it's weird that you resample such a short signal, wouldn't that introduce ringing artifacts?

This is helpful advice. I think in most cases the simulation is going to get downsampled, but it's at a relatively high frequency, so why do you call it a `short signal`. Do you agree that downsampling is still necessary? I think the goal is how to get two signal of the same number of samples while introducing minimal noise.

Reply:
I need to ponder over this whole procedure, get back to you in a couple of hours :)